crml/nft/src/lib.rs

```diff
@@ -974,6 +999,152 @@ decl_module! {
+				marketplace_id,
+			};
+			Offers::<T>::insert(offer_id, new_offer);
+			NextOfferId::mutate(|i| *i += 1);
+
+			Self::deposit_event(RawEvent::OfferMade(offer_id, amount, asset_id, marketplace_id, origin));
+			Ok(())
+		}
+
+		/// Cancels an offer on a token
+		/// Caller must be the offer buyer
+		#[weight = 0]
+		fn cancel_offer (
+			origin,
+			offer_id: OfferId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			if let Some(offer) = Self::offers(offer_id) {
+				ensure!(offer.buyer == origin, Error::<T>::NotBuyer);
+				T::MultiCurrency::unreserve(&origin, offer.asset_id, offer.amount);
+				Offers::<T>::remove(offer_id);
+				TokenOffers::remove(offer.token_id);
+				Self::deposit_event(RawEvent::OfferCancelled(offer_id));
+				Ok(())
+			} else {
+				Err(Error::<T>::InvalidOffer.into())
+			}
+		}
+
+		/// Accepts an offer on a token
+		/// Caller must be token owner
+		#[weight = 0]
+		fn accept_offer (
+			origin,
+			offer_id: OfferId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			if let Some(offer) = Self::offers(offer_id) {
+				let token_id = offer.token_id;
+				ensure!(Self::token_owner((token_id.0, token_id.1), token_id.2) == origin, Error::<T>::NoPermission);
+
+				let royalties_schedule = Self::check_bundle_royalties(&vec![token_id], offer.marketplace_id)?;
+				let for_royalties = royalties_schedule.calculate_total_entitlement() * offer.amount;
+				let mut for_seller = offer.amount;
+
+				// do royalty payments
+				if !for_royalties.is_zero() {
+					let entitlements = royalties_schedule.entitlements.clone();
+					for (who, entitlement) in entitlements.into_iter() {
+						let royalty = entitlement * offer.amount;
+						let _ = T::MultiCurrency::repatriate_reserved(&offer.buyer, offer.asset_id, &who, royalty)?;
+						for_seller -= royalty;
+					}
+				}
+
+				let seller_balance = T::MultiCurrency::free_balance(&origin, offer.asset_id);
+				let _ = T::MultiCurrency::repatriate_reserved(&offer.buyer, offer.asset_id, &origin, for_seller)?;
+
+				// The implementation of `repatriate_reserved` may take less than the required amount and succeed
+				// this should not happen but could for reasons outside the control of this module
+				ensure!(
+					T::MultiCurrency::free_balance(&origin, offer.asset_id)
+						>= seller_balance.saturating_add(for_seller),
+					Error::<T>::InternalPayment
+				);
+				Self::do_transfer_unchecked(token_id.0, token_id.1, &vec![token_id.2], &origin, &offer.buyer)?;
+
+				// Clean storage
+				Offers::<T>::remove(offer_id);
+				let mut token_offers = Self::token_offers(token_id);
+				token_offers.retain(|&x| x != offer_id);
+				TokenOffers::insert(token_id, token_offers);
+
+				Self::deposit_event(RawEvent::OfferAccepted(offer_id));
+				Ok(())
+			} else {
+				Err(Error::<T>::InvalidOffer.into())
+			}
+		}
+
+		/// Rejects an offer on a token
+		/// Caller must be token owner
+		#[weight = 0]
+		fn reject_offer (
```

Comment:
not sure this is needed so would leave out
can set the offer fee/weight higher to deter bad offers